Translate EMotionFX code to use AZStd::vector instead of MCore::Array

burelc-amzn opened the pull request.

lumberyard-employee-dm reviewed (commented):
@burelc-amzn It looks like you didn't tag anyone on this PR.
So no-one is seeing it.

moraaar commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/GLActor.cpp:
```diff
@@ -121,34 +106,35 @@ namespace RenderGL
-                mVertexBuffers[a][n] = nullptr;
-                mIndexBuffers [a][n] = nullptr;
-            }
+            vertexBuffers.resize(numGeometryLODLevels);
+            AZStd::fill(begin(vertexBuffers), end(vertexBuffers), nullptr);
```
this can be done in 1 line:
`vertexBuffers.resize(numGeometryLODLevels, nullptr);`

same with `indexBuffers`

moraaar commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/GLActor.cpp:
```diff
@@ -64,42 +55,36 @@ namespace RenderGL
-            const uint32 numIndexBuffers = mIndexBuffers[a].GetLength();
-            for (i = 0; i < numIndexBuffers; ++i)
+        }
+        for (AZStd::vector<IndexBuffer*>& indexBuffers : mIndexBuffers)
+        {
+            for (IndexBuffer* indexBuffer : indexBuffers)
             {
-                delete mIndexBuffers[a][i];
+                delete indexBuffer;
             }
         }
 
         // delete all materials
-        const uint32 numLOD = mMaterials.GetLength();
-        for (uint32 l = 0; l < numLOD; l++)
+        for (AZStd::vector<MaterialPrimitives*>& materialsPerLod : mMaterials)
         {
-            const uint32 numMaterials = mMaterials[l].GetLength();
-            for (uint32 n = 0; n < numMaterials; n++)
+            for (MaterialPrimitives* materialPrimitives : materialsPerLod)
             {
-                delete mMaterials[l][n]->mMaterial;
-                delete mMaterials[l][n];
+                delete materialPrimitives->mMaterial;
+                delete materialPrimitives;
             }
         }
     }
```
The function is actually not clearing `mVertexBuffers`, `mIndexBuffers` and `mMaterials`, leaving the elements with deleted pointers. I guess since it was working before it's not necessary?

burelc-amzn replied:
Yeah, you're right. `Cleanup()` is called from the destructor, so in that case it doesn't matter if the vector is cleared or not, as the vector's destructor itself will do that. `Cleanup()` is also called in case the initialization of the object failed, in which case (probably) none of its methods are ever called. So since it has been working this way for a while, I don't see the need to change it as part of this.

moraaar commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/GLRenderUtil.cpp:
```diff
@@ -139,10 +133,10 @@ namespace RenderGL
-        for (uint32 i = 0; i < MAX_LINE_VERTEXBUFFERS; ++i)
+        for (VertexBuffer*& lineVertexBuffer : mLineVertexBuffers)
         {
-            delete mLineVertexBuffers[i];
-            mLineVertexBuffers[i] = nullptr;
+            delete lineVertexBuffer;
+            lineVertexBuffer = nullptr;
         }
```
Similarly, `mLineVertexBuffers` is never cleared.

amzn-rhhong reviewed (approved):
Good one!

moraaar commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/Material.h:
```diff
@@ -28,18 +28,18 @@ namespace RenderGL
             mNumVertices  = 0;
 
-            mNodeIndex     = MCORE_INVALIDINDEX32;
+            mNodeIndex     = InvalidIndex;
             mMaterialIndex = MCORE_INVALIDINDEX32;
```
question: is `MCORE_INVALIDINDEX32` meant to still be used?

burelc-amzn replied:
Yes. Unfortunately that is still a little messy. The goal here was to only change variables/members to `size_t` if they are used as an index to some other thing which is now stored in an `AZStd::vector` instead of an `MCore::Array`. `mNodeIndex` here is being changed to a `size_t` because it represents an index to be passed to `EMotionFX::Skeleton::GetNode(size_t)`. `mMaterialIndex` does not, so it stays as a `uint32`, and its invalid index value has not changed.

moraaar reviewed (approved):
Reviewed cloth files. The rest I looked closer the first half and faster the second half :)

moraaar commented:
Great work replacing everything with AZStd::vector!

amzn-rhhong commented on Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp:
```diff
@@ -288,7 +288,7 @@ namespace AZ
                         {
                             // The skinned mesh lod gets a unique morph for each meta, since each one has unique min/max delta values to use for decompression
-                            AZStd::string morphString = AZStd::string::format("%s_Lod%u_Morph_%s", fullFileName.c_str(), lodIndex, metaData.m_meshNodeName.c_str());
+                            AZStd::string morphString = AZStd::string::format("%s_Lod%zu_Morph_%s", fullFileName.c_str(), lodIndex, metaData.m_meshNodeName.c_str());
```
can be const

burelc-amzn replied:
Fixed.

amzn-rhhong commented on Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/AtomActorInstance.cpp:
```diff
@@ -558,7 +557,7 @@ namespace AZ
                 // Update the morph weights for every lod. This does not mean they will all be dispatched, but they will all have up to date weights
                 // TODO: once culling is hooked up such that EMotionFX and Atom are always in sync about which lod to update, only update the currently visible lods [ATOM-13564]
-                for (uint32_t lodIndex = 0; lodIndex < m_actorInstance->GetActor()->GetNumLODLevels(); ++lodIndex)
+                for (uint32_t lodIndex = 0; lodIndex < aznumeric_cast<uint32_t>(m_actorInstance->GetActor()->GetNumLODLevels()); ++lodIndex)
```
perhaps have the numeric cast done outside the loop

burelc-amzn replied:
Fixed.

amzn-rhhong commented on Gems/EMotionFX/Code/EMotionFX/CommandSystem/Source/MotionCommands.cpp:
```diff
@@ -1282,19 +1281,17 @@ namespace CommandSystem
-            const uint32 numMotionSets = EMotionFX::GetMotionManager().GetNumMotionSets();
-            for (uint32 j = 0; j < numMotionSets; ++j)
+            const size_t numMotionSets = EMotionFX::GetMotionManager().GetNumMotionSets();
+            for (size_t j = 0; j < numMotionSets; ++j)
```
nit - bit weird it's still using j here.

burelc-amzn replied:
Fixed.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/GLRenderUtil.cpp:
```diff
@@ -139,10 +133,10 @@ namespace RenderGL
     void GLRenderUtil::CleanUp()
     {
-        for (uint32 i = 0; i < MAX_LINE_VERTEXBUFFERS; ++i)
+        for (VertexBuffer*& lineVertexBuffer : mLineVertexBuffers)
```
Is this working with mLineVertexBuffers being a pointer?

burelc-amzn replied:
`mLineVertexBuffers` is declared as:
```cpp
VertexBuffer*               mLineVertexBuffers[2]{};
```
That is a C-style array of `VertexBuffer*`s. C++11 range-for loops work on c-style arrays.

We could of course rewrite this to use an `AZStd::array<VertexBuffer*, 2>`, the usage of that member would be the same.

amzn-jillich replied:
It is fine, hopefully the code will be gone in a bit anyway.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Rendering/OpenGL2/Source/GLSLShader.cpp:
```diff
@@ -263,8 +250,8 @@ namespace RenderGL
     {
-        const uint32 index = FindAttributeIndex(name);
-        if (index == MCORE_INVALIDINDEX32)
+        const size_t index = FindAttributeIndex(name);
+        if (index == InvalidIndex)
```
Is this automatically dereferenced to the correct type? InvalidaIndex32/64 etc? I am a bit scared about that one as I remember we had some cases of using a size_t while still using the InvalidIndex32 - which is bad of course, but this might break?

burelc-amzn replied:
`InvalidIndex` is the invalid index for `size_t`:
```cpp
constexpr size_t InvalidIndex = 0xffffffffffffffff;
```
So it will be the right value for whatever size `size_t` is on the current platform.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Actor.cpp:
```diff
@@ -259,45 +241,40 @@ namespace EMotionFX
     {
         // for all LODs
-        for (uint32 i = 0; i < mMaterials.GetLength(); ++i)
+        for (AZStd::vector<Material*>& mMaterial : mMaterials)
```
mMaterial -> materialsPerLod

burelc-amzn replied:
Fixed.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Actor.cpp:
```diff
@@ -259,45 +241,40 @@ namespace EMotionFX
+            for (Material* m : mMaterial)
             {
-                mMaterials[i][m]->Destroy();
+                m->Destroy();
```
m->material

burelc-amzn replied:
Fixed.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/AnimGraphGameControllerSettings.h:
```diff
@@ -165,6 +163,6 @@ namespace EMotionFX
     private:
         AZStd::vector<Preset*>   m_presets;
-        AZ::u32                  m_activePresetIndex;
+        AZ::u64                  m_activePresetIndex;
```
What were the rules to change these automatically? Just curious why this got changed to a 64-bit int.

burelc-amzn replied:
This variable represents an index into a vector. Indexes are of type `size_t`. But this variable is also serialized. Largest `size_t` we support is 64 bits, so this variable became a `AZ::u64`.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/AnimGraphInstance.cpp:
```diff
@@ -283,28 +280,27 @@ namespace EMotionFX
         // add the remaining parameters
-        const uint32 startIndex = mParamValues.GetLength();
-        for (int32 i = 0; i < numToAdd; ++i)
+        const size_t startIndex = mParamValues.size();
+        for (ptrdiff_t i = 0; i < numToAdd; ++i)
```
Why ptrdiff_t? Do we have more instances like that?

burelc-amzn replied:
It is a `ptrdiff_t` because that is the signed version of a `size_t`.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Importer/ActorFileFormat.h:
```diff
@@ -99,6 +99,24 @@ namespace EMotionFX
         };
 
+        // (aligned)
+        struct Actor_Info4
+        {
+            uint64  mNumLODs;               // the number of level of details
+            uint64  mMotionExtractionNodeIndex;// the node number of the trajectory node used for motion extraction
+            uint64  mRetargetRootNodeIndex; // the retargeting root node index, most likely pointing to the hip or pelvis or MCORE_INVALIDINDEX32 when not set
+            uint8   mUnitType;              // maps to EMotionFX::EUnitType
+            uint8   mExporterHighVersion;
+            uint8   mExporterLowVersion;
+            uint8   mOptimizeSkeleton;
+
+            // followed by:
+            // string : source application (e.g. "3ds Max 2011", "Maya 2011")
+            // string : original filename of the 3dsMax/Maya file
+            // string : compilation date of the exporter
+            // string : the name of the actor
+        };
```
Do we need this really?? Can't we just use the already existing version and convert the values to 64-bit?

burelc-amzn replied:
I can try to remove this. I think I was probably being pedantic, that the `Actor` type now stores up to 64bit LODs, and the `Skeleton` has up to 2^64-1 nodes, so the motion extraction index and retarget root node index also have to support those index values. But if someone is making an actor with more than 4294967295 joints, we probably have other problems.

amzn-jillich replied:
Haha yeah ;) Ok, let's remove it and keep the old version.

burelc-amzn replied:
I've removed the new Actor version from this PR.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp:
```diff
@@ -769,37 +756,33 @@ namespace EMotionFX
         {
             outVertexCounts[0] = GetNumVertices();
-            return aznumeric_cast<uint32>(maxInfluences);
+            return 0;
```
Why are we just returning 0 here while we returned the maxInfluences before?

burelc-amzn replied:
The old code was this:

```cpp
  1 uint32 Mesh::CalcMaxNumInfluences(AZStd::vector<uint32>& outVertexCounts) const
  2 {
  3     size_t maxInfluences = 0;
  4
  5     // Reset values.
  6     outVertexCounts.resize(CalcMaxNumInfluences() + 1);
  7     for (size_t j = 0; j < outVertexCounts.size(); ++j)
  8     {
  9         outVertexCounts[j] = 0;
 10     }
 11
 12     // Does the mesh have a skinning layer? If no we can quit directly as this means there are only unskinned vertic    es.
 13     SkinningInfoVertexAttributeLayer* skinningLayer = (SkinningInfoVertexAttributeLayer*)FindSharedVertexAttributeLa    yer(SkinningInfoVertexAttributeLayer::TYPE_ID);
 14     if (!skinningLayer)
 15     {
 16         outVertexCounts[0] = GetNumVertices();
 17         return aznumeric_cast<uint32>(maxInfluences);
 18     }
```

You can see on line 3 that `maxInfluences` is initialized to `0`, and never modified before the return on line 17. So if there's no skinning layer, it will return `0`. This seems correct to me.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp:
```diff
@@ -826,22 +809,21 @@ namespace EMotionFX
         // check the types of all vertex attribute layers
-        const uint32 numAttributes = mVertexAttributes.GetLength();
-        for (uint32 i = 0; i < numAttributes; ++i)
+        for (auto *vertexAttribute : mVertexAttributes)
```
auto * -> auto*

burelc-amzn replied:
Fixed.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/MorphTargetStandard.cpp:
```diff
@@ -169,28 +169,27 @@ namespace EMotionFX
         // calculate the new transformations for all nodes of this morph target
-        const uint32 numTransforms = mTransforms.GetLength();
-        for (uint32 i = 0; i < numTransforms; ++i)
+        for (const Transformation& mTransform : mTransforms)
```
mTransform -> transform
This seems to have happened on multiple places.

burelc-amzn replied:
I fixed all the places where the loop variable was named after the member variable that it was iterating over. See 
8a286ee.

amzn-jillich reviewed (commented):
Great changes but also scary as this is basically impossible to review due to the amount of changes.
Part 1 (got till MotionInstance.h, but not even comprehensively and more browsing through things)

burelc-amzn commented on Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp:
```diff
@@ -57,8 +57,8 @@ namespace
 
-    const size_t LinearSkinningFloatsPerBone = 12;
-    const size_t DualQuaternionSkinningFloatsPerBone = 8;
+    const uint32_t LinearSkinningFloatsPerBone = 12;
+    const uint32_t DualQuaternionSkinningFloatsPerBone = 8;
```
I used `uint32_t` so that the template call to `AZStd::GetMin<uint32_t>` on line 214 works.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Source/Recorder.h:
```diff
@@ -193,50 +193,66 @@ namespace EMotionFX
+            size_t                                      mByteOffset = 0;
+            size_t                                      mNumBytes = 0;
+            AZStd::vector<AnimGraphAnimObjectInfo>      mObjectInfos{};
+            AZStd::vector<AZStd::unique_ptr<MCore::Attribute>>             mParameterValues{};
```
Did you test the recorder? We're cloning the attributes when recording and it should be safe, but let's give that a testing round as the recorder usually is pretty sensitive.

burelc-amzn replied:
I did some basic tests with the recorder and it seems to be working.

amzn-jillich commented on Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/AnimGraph/NodeGraphWidget.cpp:
```diff
@@ -1095,7 +1087,7 @@ namespace EMStudio
 
                 // get the information from the current mouse position
-                uint32      newTargetPortNr;
+                AZ::u16     newTargetPortNr;
```
Why AZ::u16 here? Can we align to a standard with this change now that we're changing it anyway?

burelc-amzn replied:
That variable is passed to `NodeGraph::FindPort`, which calls `GraphNode::FindPort(int32 x, int32 y, AZ::u16* outPortNr, bool* outIsInputPort, bool includeInputPorts)`. The signature of that method was changed to match the type that that function uses internally. That snippet is this:
```cpp
  1003 const AZ::u16 numInputPorts = aznumeric_caster(mInputPorts.size());
  1004 for (AZ::u16 i = 0; i < numInputPorts; ++i)
  1005 {
  1006     QRect rect = CalcInputPortRect(i);
  1007     if (rect.contains(QPoint(x, y)))
  1008     {
  1009         *outPortNr      = i;
  1010         *outIsInputPort = true;
  1011         return &mInputPorts[i];
  1012     }
  1013 }
```

It was already using `AZ::u16`, so I just changed the parameter type to match.

burelc-amzn replied:
If we want to do a separate pass to unify our `uint32` and `AZ::u32` and `uint32_t`, I'm all for it, but I think that's separate from this change.

amzn-jillich replied:
Sounds good

amzn-jillich commented on Gems/EMotionFX/Code/MCore/Source/AlignedArray.h:
```diff
@@ -18,7 +18,7 @@ namespace MCore
-     * It can theoretically store 4294967296 items (maximum uint32 value).
+     * It can theoretically store 18446744073709551614 items (maximum size_t value - 1 for the invalid index).
      */
     template <typename T, uint32 alignment>
     class AlignedArray
```
We can also get rid of the aligned array directly, or do it in a second pass.  The AZStd::vector<> is using aligned memory allocations already, so there is no use for this class anymore, too.

burelc-amzn replied:
I say we get this in first, then do a separate round for `AlignedArray`.

amzn-jillich reviewed (commented):
Phew, went through most of it :). Great stuff, can you get the feedback in and after that we'll hand it over to QA for a testing round?

burelc-amzn commented on Gems/EMotionFX/Code/EMotionFX/Source/Pose.cpp:
```diff
@@ -613,11 +589,10 @@ namespace EMotionFX
+            for (size_t i = 0; i < numNodes; ++i)
             {
-                nodeNr = actorInstance->GetEnabledNode(i);
+                const uint16 nodeNr = actorInstance->GetEnabledNode(i);
```
Generally I'm using whatever type is used already by whatever gem was updated. I don't see any `uint16_t`, but I know that Atom and NvCloth use `uint32_t`, whereas EMotionFX uses `uint32`.

burelc-amzn commented on Gems/EMotionFX/Code/EMotionFX/Source/MotionSet.cpp:
```diff
@@ -734,38 +721,25 @@ namespace EMotionFX
-        }
-        return nullptr;
+            return !motionSet->GetIsOwnedByRuntime() && iter-- == 0;
+        });
```
I used std algorithms to describe what the code is doing rather than how it does it. This code is finding an element in a vector, so `find_if` makes sense. It has the added logic of not finding the first match, but the `index`th match.

burelc-amzn commented:
I resolved other merge conflicts and addressed the issues Roman mentioned, but I still need to revert the new version to the Actor binary file format.
